README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# different.tan

<!-- badges: start -->
<!-- badges: end -->

R code for the analysis of vehicle emissions remote sensing systems
(VERSS).

> In development package

## Installation

You can install the development version of different.tan from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("karlropkins/different.tan")
```

## Contributing

**different.tan** is an in-development package, and contributions are
very welcome. We are using tidyverse contribution guidance regarding
code of conduct but are not strict as them regarding coding style…
[Contributions](https://karlropkins.github.io/different.tan/CONTRIBUTING.html)


NEWS

different.tan 0.1.1 [2022-11-24]
=====================================
  
* posted on github 
* updated citation information 


different.tan 0.1.0 [2022-09-03]
=====================================

* First package compile for public version 
  of package; rebuilding using 'devtools' 
  conventions. 
* (earlier packages archived; but intending 
  re-adding all code we are able to share).
* Re-Adding plot_Pareto and plot_ParetoEnrichment. 
* Added import_featdata for route download 
  of data from the FEAT data archive. 
  

        









 





R/different.tan-package.R
##################################################
#' different.tan
##################################################
#'
#' different.tan: R code for the analysis of 
#' vehicle emissions remote sensing system (VERSS) 
#' data 
#'
#' @section different.tan:
#' \code{different.tan} was initially developed as part 
#' of work on a series of projects collecting and 
#' analyzing EDAR data, but many also be of interest to those using other similar (VERSS) methods.
#'
#' The name 'different.tan' refers to Don Stedman, a pioneer 
#' in field, who died in 2016: 
#' \href{https://magazine-archive.du.edu/campus-community/chemistry-professor-donald-stedman-dies-lung-cancer/}{[DON]}
#' 
#' @author Karl Ropkins, Francis Pope, David Beddows, Omid Ghaffarpasand 
#' 
#' @references 
#' We gratefully acknowledge the huge contributions of the R Core Team 
#' and numerous contributors in developing and maintaining R:
#' 
#' R Development Core Team (2011). R: A language and environment for 
#' statistical computing. R Foundation for Statistical Computing, 
#' Vienna, Austria. ISBN 3-900051-07-0, URL http://www.R-project.org/.
#' 
#' Ropkins, K., DeFries, T.H., Pope, F., Green, D.C., Kemper, J., Kishan, S., 
#' Fuller, G.W., Li, H., Sidebottom, J., Crilley, L.R. and Kramer, L. (2017). 
#' Evaluation of EDAR vehicle emissions remote sensing technology. Science of 
#' the Total Environment, 609, pp.1464-1474. https://doi.org/10.1016/j.scitotenv.2017.07.137 
#' 
#' Ghaffarpasand, O., Beddows, D.C., Ropkins, K. and Pope, F.D. (2020). Real-world 
#' assessment of vehicle air pollutant emissions subset by vehicle type, fuel and 
#' EURO class: New findings from the recent UK EDAR field campaigns, and implications 
#' for emissions restricted zones. Science of The Total Environment, 734, p.139416.
#' https://doi.org/10.1016/j.scitotenv.2020.139416 
#' 
#' Ghaffarpasand, O., Ropkins, K., Beddows, D.C. and Pope, F.D. (2022). Detecting high 
#' emitting vehicle subsets using emission remote sensing systems. Science of The Total 
#' Environment, p.159814.https://doi.org/10.1016/j.scitotenv.2022.159814
#' 
#' @docType package
#' @name different.tan
NULL


R/plot_ParetoEnrichment.R
############################################
#' @title Pareto Enrichment Plots
############################################

#' @name plot_ParetoEnrichment
#' @description [to be added]
#' @param x [to be added]
#' @param y [to be added]
#' @param cut [to be added]
#' @param data [to be added]
#' @param ... [to be added]
#' @param min.bin [to be added]
#' @param drop.nas [to be added]
#' @param order [to be added]
#' @param level.nmax [to be added]

#splatted function
#' @export
plot_ParetoEnrichment <- 
function (x, y=NULL, cut=0.8, data = NULL, ..., 
                                  min.bin=1, drop.nas = TRUE, order=FALSE, level.nmax=NULL) 
{
    
    #to do
    #check which formals are used 
    #think about group and cond
    #think about cut currently only expecting one...
    #think about key
    #tidy code so we can flip axes
    #tidy code so we can update grid and line elements
    #think about other options to display results
    #overlapping histograms?
    
    
    #setup
    argnames <- names(as.list(match.call(expand.dots = TRUE)[-1]))
    arguments <- as.list(match.call()[-1])
    extra.args <- list(...)
    
    #names
    #currently no groups or cond
    x.name <- as.character(arguments)[argnames == "x"]
    y.name <- as.character(arguments)[argnames == "y"]
    
    #name updates
    y.name <- paste("cutPareto(", y.name, ", ", cut, ") Enrichment", sep="")
    
    #currently no key
    #    #key
    #    extra.args <- do.call(listLoad, listUpdate(extra.args, list(load = "key")))
    #    key <- extra.args$key
    #    if(is.null(key)) key=TRUE
    #    extra.args <- extra.args[!names(extra.args) %in% c("key")]
    
    #data source
    if (!is.null(data)) 
        data <- as.data.frame(data)
    env <- parent.frame()
    #currently groups and cond not checked for when making df
    df <- list(x = eval(substitute(x), data, env), y = eval(substitute(y), 
                                                            data, env))
    ref <- lapply(df, length)
    df <- as.data.frame(df[ref > 0])
    
    calc.df <- calcPareto(df$y, output="all") 
    df <- cbind(df[calc.df$ind,], calc.df)
    
    #return(df)
    
    #note currently flipping axes
    #might want to think about this
    
    #set default col
    
    col <- colHandler(z=10, zlim=c(1, 10), col.regions="Spectral")
    
    plot.args <- listUpdate(list(x=x~y, data=df, xlab=y.name, ylab=x.name, origin=0, col=col), extra.args)
    plt <- do.call(xyplot, plot.args)
    
    #rework outputs
    #this assumes 
    #above x and y structure - y factor x numeric
    #cut is currently in panel but assuming only one panel...
    #   so we can look at multiple panel
    #   so could simplify code if we decide definitely not to do this
    
    #doing this is way so cut is in-panel not all data  
    
    min.max <- c() 
    rt <- rep(0, length(levels(plt$panel.args[[1]]$y)))
    for(i in 1:length(plt$panel.args)){
        all.data <- data.frame(x=plt$panel.args[[i]]$x, y=plt$panel.args[[i]]$y) 
        cum.sum <- cumsum(all.data$x)
        cum.sum <- cum.sum/max(cum.sum, na.rm=TRUE)
        hi.cut <- subset(all.data, cum.sum <= cut)
        
        y <- summary(all.data$y, maxsum=length(levels(all.data$y)))
        y[y < min.bin] <- NA
        
        y <- y/sum(y, na.rm=TRUE)
        #levels should be same
        y2 <- summary(hi.cut$y, maxsum=length(levels(all.data$y)))
        y2 <- y2/sum(y2, na.rm=TRUE)
        y <- y2/y
        
        plt$panel.args[[i]]$x <- as.vector(y)
        min.max <- c(min(c(min.max, as.vector(y)), na.rm=TRUE), max(c(min.max, as.vector(y)), na.rm=TRUE))
        plt$panel.args[[i]]$y <- factor(names(y), levels=levels(plt$panel.args[[i]]$y))
        y[is.na(y)] <- 0
        rt <- rt+y
    }
    
    border <- (min.max[2]-min.max[1])*0.05
    plt$x.limits <- min.max + c(-border, border)
    if("origin" %in% names(plt$panel.args.common))
        plt$x.limits[1] <- 0-border
    plt$panel.args.common$horizontal <- TRUE
    
    if(order){
        plt$y.limits <- plt$y.limits[order(rt)]
        for(i in 1:length(plt$panel.args)){
            plt$panel.args[[i]]$x <- plt$panel.args[[i]]$x[order(rt)]
            plt$panel.args[[i]]$y <- as.character(plt$panel.args[[i]]$y[order(rt)])
            plt$panel.args[[i]]$y <- factor(plt$panel.args[[i]]$y, levels=plt$y.limits)
        }
        rt <- rt[order(rt)]
    }
    if(drop.nas & any(rt==0)){
        
        #this might not completely work
        #if enrichment is zero not NA...
        #if rt contains NAs...
        
        plt$y.limits <- plt$y.limits[rt!=0]
        for(i in 1:length(plt$panel.args)){
            plt$panel.args[[i]]$x <- plt$panel.args[[i]]$x[rt!=0]
            plt$panel.args[[i]]$y <- as.character(plt$panel.args[[i]]$y[rt!=0])
            plt$panel.args[[i]]$y <- factor(plt$panel.args[[i]]$y, levels=plt$y.limits)
        }
        rt<-rt[rt!=0]
    }
    
    if(is.numeric(level.nmax)){
        temp <- if(order) rev(rev(1:length(plt$y.limits))[1:level.nmax]) else 1:level.nmax
        if(length(plt$y.limits > level.nmax)){
            plt$y.limits <- plt$y.limits[temp]
            for(i in 1:length(plt$panel.args)){
                plt$panel.args[[i]]$x <- plt$panel.args[[i]]$x[temp]
                plt$panel.args[[i]]$y <- as.character(plt$panel.args[[i]]$y[temp])
                plt$panel.args[[i]]$y <- factor(plt$panel.args[[i]]$y, levels=plt$y.limits)
            }
            #rt<-rt[rt!=0]
            if(!order)
                warning(paste("In ParetoEnrichment(): Note, just showing first ", 
                   level.nmax, " levels\n\tReplot with order=TRUE if you want highest", sep=""), 
                   call.=FALSE)
            
        }
    }
    
    plt$panel <- function(...){
        panel.grid(-1,-1)
        panel.barchart(...)
        panel.abline(v=1,col="red",lty=2)                       
    }
    if(length(plt$panel.args[[1]]$x) > 40)
        warning(paste("In ParetoEnrichment(): Large '", x.name, 
            "' level number? \n\tMaybe refactorise, or apply min.bin, order and/or level.nmax?", 
            sep=""), call.=FALSE)
    
    plt
    
}





R/plot_Pareto.R
############################################
#' @title plot_Pareto
############################################

#' @name plot_Pareto 
#' @description  In-development plots functions intended for use with VERSS data. 
#' @param y (numeric vector), measurements to plotted as PartoPlot Y-axis terms. In VERSS 
#' analysis these are typically emission measurements from passing vehicles, but can be 
#' any data-series where the frequency of occurrence of \code{y} is of interest. 
#' @param groups (optional) sub-sampling term for grouping measurements within a plot.  
#' @param cond (optional) sub-sampling term for plotting data as separate panels. 
#' @param data (optional, data.frame or similar), source of \code{y}, \code{groups},
#' \code{cond}, etc.
#' @param x (optional, number vector), by default, the index of \code{y} values, lowest 
#' to highest.
#' @param ... Passed to lattice and loa plot functions (and used if possible).
#' @param panel lattice plot argument
#' @param drop.nas Return NAs before plotting.
#' @param Pareto (Character) Pareto plot type, options: 'Pareto' standard Pareto; 
#' 'cum.Pareto' cumulative Pareto; 'normal.cum.Pareto' cumulative Pareto expressed 
#' as a proportion; and, 'percent.cum.Pareto' cumulative Pareto expressed 
#' as a percentage.
#' @returns \code{lattice} object plot. 


#splatted function
#version v0.1.2 kr 
#' @export
plot_Pareto <- 
function (y, groups = NULL, cond = NULL, data = NULL, x = NULL,  
          ..., panel=panel.ParetoPlot, drop.nas = TRUE, 
          Pareto = "normal.cum.Pareto") 
{
    
    #to do
    #check args are aligned 
    #e.g. does drop.nas link into drop.unused.levels
    #     and how should it do this???
    
    
    #setup
    argnames <- names(as.list(match.call(expand.dots = TRUE)[-1]))
    arguments <- as.list(match.call()[-1])
    extra.args <- list(...)
    
    #names
    x.name <- as.character(arguments)[argnames == "x"]
    y.name <- as.character(arguments)[argnames == "y"]
    groups.name <- as.character(arguments)[argnames == "groups"]
    cond.name <- as.character(arguments)[argnames == "cond"]
    
    #key
    extra.args <- do.call(listLoad, listUpdate(extra.args, list(load = "key")))
    key <- extra.args$key
    if(is.null(key)) key=TRUE
    extra.args <- extra.args[!names(extra.args) %in% c("key")]
    
    #data source
    if (!is.null(data)) 
        data <- as.data.frame(data)
    env <- parent.frame()
    df <- list(x = eval(substitute(x), data, env), 
               y = eval(substitute(y), data, env), 
               groups = eval(substitute(groups), data, env), 
               cond = eval(substitute(cond), data, env))
    ref <- lapply(df, length)
    df <- as.data.frame(df[ref > 0])
    
    calc.df <- calcPareto(df$y, output="all") 
    df <- cbind(df[calc.df$ind,], calc.df)
    df$y <- df[,Pareto]
    y.name <- paste(Pareto, "(", y.name, ")", sep="")
    
    #add x if not supplied
    if (is.null(df$x)){
        df$x <- c(1:length(df$y))
        if(length(x.name)<1){
            x.name <- "count"
        }
    }
    
    form <- "y~x"
    form <- as.formula(form)
    grps <- if (is.null(df$groups)) 1 else 
        if (is.factor(df$groups)) levels(df$groups) else 
            sort(unique((df$groups)))
    temp <- colHandler(1:20, col.regions = "Spectral")
    temp <- listUpdate(listUpdate(list(col.regions = temp), extra.args), 
                       list(z = 1:length(grps)), ignore = c("zlim"))
    col <- if (length(temp$z) < 2) {
        temp$z <- c(1, 2)
        do.call(colHandler, temp)[2]
    }
    else do.call(colHandler, temp)
    
    plot.list <- list(x=form, data=df, xlab=x.name, ylab=y.name, col=col, 
                      drop.unused.levels=drop.nas, method=Pareto,
                      panel = panel, grid=TRUE, parent=TRUE, line=TRUE, 
                      shadow = TRUE)
    if("groups" %in% names(df))
        plot.list$groups <- df$groups
    plot.list <- listUpdate(plot.list, extra.args)
    
    for (i in c("grid", "line", "shadow", "parent")) 
      plot.list <- do.call(listLoad, listUpdate(plot.list, list(load = i)))
    
    plt <- do.call(xyplot, plot.list)
    
    if("cond" %in% names(df)){
        plt$panel.args.common$cond <- df$cond
        cnds <- if (is.factor(df$cond)) levels(df$cond) else 
            sort(unique((df$cond)))
        if(plot.list$drop.unused.levels){
            ans <- sapply(cnds, function(x) length(df$cond[df$cond==x]))
            cnds <- cnds[c(ans>0)]
        } 
        for(i in 1:length(cnds)){
            plt$panel.args[[i]] <- plt$panel.args[[1]]
            plt$panel.args[[i]]$local.y <- rep(0, length(plt$panel.args[[i]]$y))
            temp <- plt$panel.args[[i]]$y
            if(grepl("cum", tolower(plt$panel.args.common$method)))
                temp <- c(temp[1], diff(temp))
            plt$panel.args[[i]]$local.y[df$cond==cnds[i]] <- temp[df$cond==cnds[i]]
        }
        plt$condlevels <- list(cnds)
        plt$index.cond <- list(1:length(cnds))
        #strip could be set up user
        if(is.logical(plt$strip) && !plt$strip)
            plt$strip <- "strip.default"
    } else {
        plt$panel.args[[1]]$local.y <- plt$panel.args[[1]]$y
        if(grepl("cum", tolower(plt$panel.args.common$method)))
            plt$panel.args[[1]]$local.y <- c(plt$panel.args[[1]]$local.y[1], 
                                             diff(plt$panel.args[[1]]$local.y))
    }
    
    #log method?
    plt$loa <- list(results = df)
    
    #make key if requested
    if("groups" %in% names(df) & isGood4LOA(key)){
        
        grps <- rev(levels(as.factor(df$groups)))
        col <- rev(col)
        #col already declared - line col
        
        if(isGood4LOA(plt$panel.args.common$shadow)){
            temp <- listUpdate(listUpdate(list(border=FALSE), extra.args), 
                               plt$panel.args.common$shadow, ignore="groups")
            if(!"alpha" %in% names(temp))
                temp$alpha <- 0.5
            shadow.col <- do.call(colHandler, listUpdate(temp, list(z=1:length(col), 
                                                                    col.regions=col)))
            
        }
        
        my.key <- list(space="right", title=if ("main" %in% names(key)) key$main else groups.name,
                       cex.title=1, adj=0, between=1)
        
        if(isGood4LOA(plt$panel.args.common$line) & !isGood4LOA(plt$panel.args.common$shadow)){
            my.key$lines <- list(col=col, size=2) 
            if(!is.null(plt$panel.args.common$lty)) my.key$lines$lty <- plt$panel.args.common$lty
            if(!is.null(plt$panel.args.common$lwd)) my.key$lines$lwd <- plt$panel.args.common$lwd
        }
        if(isGood4LOA(plt$panel.args.common$shadow) & isGood4LOA(plt$panel.args.common$line))
            my.key$rect <- list(col=shadow.col, border=col, size=2) 
        if(isGood4LOA(plt$panel.args.common$shadow) & !isGood4LOA(plt$panel.args.common$line))
            my.key$rect <- list(col=shadow.col, border=FALSE, size=2) 
        my.key$text <- list(grps)
        my.key$rep <- TRUE
        
        key <- listUpdate(my.key, key)
        plt$legend <- list(right=list(fun="draw.key", args=list(key=key, draw=FALSE)))
        
        #######################
        #key needs more work
        #######################
        
        if(key$space != "right"){
            names(plt$legend)[1] <- key$space 
            if(key$space == "inside"){
                if("x" %in% names(key))
                    plt$legend$inside$args$x <- key$x
                if("y" %in% names(key))
                    plt$legend$inside$args$y <- key$y
            }
        }
        
    }
    
    plt
    
}


#local functions for now

panel.ParetoPlot <- function(...){
  
  #set up
  extra.args <- list(...)
  
  #track grid, line, shadow, parent settings
  
  #could we do this in the main plot?
  
  for (i in c("grid", "line", "shadow", "parent")) extra.args <- do.call(listLoad, 
                                                                         listUpdate(extra.args, list(load = i)))
  grid <- extra.args$grid
  line <- extra.args$line
  shadow <- extra.args$shadow
  parent <- extra.args$parent
  extra.args <- extra.args[!names(extra.args) %in% c("grid", 
                                                     "line", "shadow", "parent")]
  
  #then it would be sorted in all cases where user looks at this...?
  
  #show grid if requested
  if(isGood4LOA(grid))
    do.call(panel.grid, listUpdate(list(h=-1, v=-1), grid))
  
  #check for method
  cumsum <- if(!"method" %in% names(extra.args)){
    warning("method unknown: this may not work if data is cumsum...")
    FALSE
  } else { 
    grepl("cum", tolower(extra.args$method))
  }      
  
  #check for group and cond
  cond <- if("cond" %in% names(extra.args)) TRUE else FALSE
  groups <- if("groups" %in% names(extra.args)) TRUE else FALSE
  
  if(any(c(cond, groups)))
    if(isGood4LOA(parent)){
      temp <- listUpdate(listUpdate(list(type="l", lty=2), extra.args), parent, ignore="groups")
      if(!is.null(temp$col))
        temp$col <- temp$col[length(temp$col)]
      do.call(panel.xyplot, temp)
    }
  
  if(!groups) extra.args$groups <- rep(1, length(extra.args$y))
  grps <- levels(as.factor(extra.args$groups))
  prev.y <- rep(0, length(extra.args$y))
  
  ###################################################
  #should probably make new panel or plot for cumsum = FALSE
  #too much overplotting
  ###################################################
  
  for(i in 1:length(grps)){
    
    #think about this 
    #it does not like not dropping the NAs
    y <- rep(0, length(extra.args$local.y))
    y[extra.args$groups == grps[i] & !is.na(extra.args$groups)] <- extra.args$local.y[extra.args$groups == grps[i] & !is.na(extra.args$groups)]
    if(cumsum) y <- cumsum(y)
    #think about this
    temp <- listUpdate(listUpdate(list(border=FALSE), extra.args), shadow, ignore="groups")
    temp$x <- c(temp$x, rev(temp$x))
    temp$y <- c(y+prev.y, rev(prev.y))
    prev.y <- y + prev.y
    if(!"alpha" %in% names(temp))
      temp$alpha <- 0.3
    temp$col <- extra.args$col[i]
    temp$border <- if(isGood4LOA(line)) extra.args$col[i] else FALSE
    if(isGood4LOA(shadow) & !all(y==0))
      do.call(panel.polygon, temp)
    temp <- listUpdate(temp, line)
    temp$x <- extra.args$x
    temp$y <- prev.y
    temp$type <- "l"
    if(isGood4LOA(line) & !all(y==0))
      do.call(panel.xyplot, temp)
  }   
  
  transpose <- do.call(listLoad, listUpdate(list(load="transpose", transpose.col="red", transpose.label.format=2), extra.args))$transpose   
  if(any(c("y", "y.normal", "y.percent") %in% names(transpose))){
    borders <- current.panel.limits(unit="native")
    diffs <- c((borders$xlim[2]-borders$xlim[1])*0.05, (borders$ylim[2]-borders$ylim[1])*0.05)
    lims <- c(min(extra.args$y, na.rm=TRUE), max(extra.args$y, na.rm=TRUE))
    
    if(!"y" %in% names(transpose)){
      if("y.normal" %in% names(transpose)) transpose$y <- max(lims, na.rm=TRUE) * transpose$y.normal 
      if("y.percent" %in% names(transpose)) transpose$y <- max(lims, na.rm=TRUE)* (transpose$y.percent/100) 
    }
    for(i in transpose$y){
      temp <- extra.args$x[extra.args$y >= i][1]
      do.call(panel.lines, listUpdate(transpose, list(y=rep(i, 2), x=c(borders$xlim[1], temp))))
      do.call(panel.arrows, listUpdate(transpose, list(x0=temp, x1=temp, y0=i, y1=borders$ylim[1])))
      label <- as.character(signif(i, 4))
      if(transpose$label.format==2) label <- paste(signif((i/lims[2])*100, 2), "%", sep="")
      if(transpose$label.format==3) label <- paste(label, " (", signif((i/lims[2])*100, 2), "%)", sep="")
      do.call(panel.text, listUpdate(transpose, list(x=borders$xlim[1], y=i+diffs[2], label=label, pos=4)))
      label <- as.character(round(temp))
      if(transpose$label.format==2) label <- paste(signif((temp/max(extra.args$x, na.rm=TRUE))*100, 2), "%", sep="")
      if(transpose$label.format==3) label <- paste(label, " (", signif((temp/max(extra.args$x, na.rm=TRUE))*100, 2), "%)", sep="")
      do.call(panel.text, listUpdate(transpose, list(x=temp, y=borders$ylim[1]+(2*diffs[2]), label=label, pos=4)))
    }
  }
  
}



#local function for now...
calcPareto <- function(x, na.rm=TRUE, neg2zero=TRUE, method="normal.cum.Pareto", ..., output="ans"){
  #function calculates various Pareto terms 
  
  x <- as.numeric(x)
  
  ##########################
  #want to hold onto the 
  #original indices
  #for output != "ans"
  ##########################
  
  d <- data.frame(ind=1:length(x), Pareto=x)
  if(na.rm) d <- na.omit(d)
  #order highest first
  d <- d[rev(order(d$Pareto)),]
  
  ##############################
  #doing this after ordering 
  #assumes negs affect order of
  #the zero cases
  ############################## 
  if(neg2zero) d$Pareto[d$Pareto<0] <- 0
  temp <- d$Pareto
  temp[is.na(temp)] <- 0
  if(method=="Pareto" & output=="ans") return(d$Pareto)
  
  d$cum.Pareto <- cumsum(temp)
  if(method=="cum.Pareto" & output=="ans") return(d$cum.Pareto)
  
  d$normal.cum.Pareto <- d$cum.Pareto/max(d$cum.Pareto, na.rm=TRUE)
  if(method=="normal.cum.Pareto" & output=="ans") return(d$normal.cum.Pareto)
  
  d$percent.cum.Pareto <- d$normal.cum.Pareto * 100
  if(method=="percent.cum.Pareto" & output=="ans") return(d$percent.cum.Pareto)
  
  row.names(d) <- 1:nrow(d) 
  
  #output
  d
}


#local function for now

cutPareto <- function(x, cut=0.8, cond=NULL, ...){
  
  #this is function to chop Pareto series
  
  pareto <- calcPareto(x, output="all")
  if(!is.null(cond)) pareto$cond <- as.factor(cond[pareto$ind])
  
  ans <- lapply(cut, function(x){
    total <- pareto$cum.Pareto[length(pareto$cum.Pareto)]
    subtotal <- total * x
    temp <- subset(pareto, pareto$cum.Pareto <= subtotal)
    out <- c(total=total, cut=x, subtotal=subtotal, n=nrow(pareto), subn=nrow(temp), prop=nrow(temp)/nrow(pareto))
    if(is.null(cond)) out else c(out, summary(temp$cond))
  })
  ans <- do.call(rbind, ans)
  ans
}



